crates/qpy/src/circuits.rs

```diff
@@ -0,0 +1,2166 @@
+            key: formats::condition_types::NONE,
+            register_size: 0u16,
+            value: 0i64,
+            data: formats::ConditionData::None,
+        }),
+        Some(condition) => {
+            if condition.extract::<classical::expr::Expr>().is_ok() {
+                let expression = pack_generic_data(&condition, qpy_data)?;
+                Ok(formats::ConditionPack {
+                    key: formats::condition_types::EXPRESSION,
+                    register_size: 0u16,
+                    value: 0i64,
+                    data: formats::ConditionData::Expression(expression),
+                })
+            } else if condition.is_instance_of::<PyTuple>() {
+                let key = formats::condition_types::TWO_TUPLE;
+                let value = condition
+                    .downcast::<PyTuple>()?
+                    .get_item(1)?
+                    .extract::<i64>()?;
+                let register =
+                    dumps_register(condition.downcast::<PyTuple>()?.get_item(0)?, circuit_data)?;
+                Ok(formats::ConditionPack {
+                    key,
+                    register_size: register.len() as u16,
+                    value,
+                    data: formats::ConditionData::Register(register),
+                })
+            } else {
+                Err(PyValueError::new_err(
+                    "Expression handling not implemented for get_condition_data_from_inst",
+                ))
+            }
+        }
+    }
+}
+
+fn get_condition_data(
+    py: Python,
+    op: &PackedOperation,
+    circuit_data: &CircuitData,
+    qpy_data: &QPYWriteData,
+) -> PyResult<formats::ConditionPack> {
+    match op.view() {
+        OperationRef::Instruction(py_inst) => {
+            get_condition_data_from_inst(py, &py_inst.instruction, circuit_data, qpy_data)
+        }
+        // we assume only PyInstructions have condition data at this stage
+        _ => Ok(formats::ConditionPack {
+            key: formats::condition_types::NONE,
+            register_size: 0u16,
+            value: 0i64,
+            data: formats::ConditionData::None,
+        }),
+    }
+}
+
+fn recognize_custom_operation(
+    py: Python,
+    op: &PackedOperation,
+    name: &String,
+) -> PyResult<Option<String>> {
+    let library = py.import("qiskit.circuit.library")?;
+    let circuit_mod = py.import("qiskit.circuit")?;
+    let controlflow = py.import("qiskit.circuit.controlflow")?;
+
+    if (!library.hasattr(name)?
+        && !circuit_mod.hasattr(name)?
+        && !controlflow.hasattr(name)?
+        && name != "Clifford")
+        || name == "Gate"
+        || name == "Instruction"
+        || is_python_gate(py, op, imports::BLUEPRINT_CIRCUIT.get_bound(py))?
+    {
+        // Assign a uuid to each instance of a custom operation
+        let new_name = if !["ucrx_dg", "ucry_dg", "ucrz_dg"].contains(&op.name()) {
+            format!("{}_{}", &op.name(), Uuid::new_v4().as_simple())
```

Comment:
Is there a special meaning to storing the UUID in this case as simple vs not in the other cases?